### ISSUE TITLE
In search of an XLIFF editor

### ISSUE THREAD

**Issue by ghost** [EFForg/www-l10n#2]
Hi there,

I plan on helping out with the translations (English -> Turkish) but I haven't been able to find a good XLIFF editor for Mac OS X. I checked the [list of apps on Wikipedia](https://en.wikipedia.org/wiki/XLIFF#XLIFF_Related_Tools) but every link on that page has its own different problems. (Some are paid, some links are dead, some requires Java which I refuse to install on any of my computers.) Since I couldn't find a proper app to handle .xliff files (or rather, file) can I get some recommendations?

I bought [Xliffie](https://itunes.apple.com/us/app/xliffie/id961651406?mt=12) and [iXLIFF](https://itunes.apple.com/us/app/ixliff/id944658897?mt=12) from the Mac App Store but they didn't let me edit the file. I also downloaded Xcode hoping that it'd have its own .xliff editor since it's able to create and export .xliff files but that didn't work either. The only solution that worked was to download the trial version of an app called [Counterparts Lite](https://itunes.apple.com/us/app/counterparts-lite/id711576477?mt=12), but it costs $19 for continued use.

I'd appreciate some help.

Thanks.

**Comment by mfb** [comment]
Hi @dombili I don't have a Mac but if I can get my hands on one I'll try some things out.

You could also try the cloud-based translation tools, like memsource.com which has a free personal edition or transifex.com where I already uploaded it: https://www.transifex.com/projects/p/eff/resource/jobid35_en_trxlf/ or others..

**Comment by ghost** [comment]
Thanks for your reply, mfb.

I'm somewhat familiar with Transifex, so I tried that first after you added Turkish language, but it doesn't look like Transifex handles xliff files (at least the one you uploaded) well. Here's what I see when I try to start the translation process:

![screen shot 2015-07-31 at 07 22 37](https://cloud.githubusercontent.com/assets/3340067/9000770/3d4df3cc-3755-11e5-8d17-1f4075e619ec.png)

As you can see, some strings are properly shown, but some are just a mess. (I'm not familiar with xliff files, so I assume it doesn't suppose to do that.) For example, when I use the .xliff file I got from GitHub on the app I mentioned that works (which I only have the trial version of), strings look like this:

<img width="1155" alt="screen shot 2015-07-31 at 07 29 17" src="https://cloud.githubusercontent.com/assets/3340067/9000815/ff8882fe-3755-11e5-93ae-35dd9a488645.png">

That's much more cleaner and easy to work with. I'm sure you'd agree with me.

I can work on Transifex, but as of right now, I'd have to double the time I'm going to spend translating because of how messy the strings look. I'd really appreciate it if you can find a proper solution for Mac (to work locally) or Transifex.

Thank you.

**Comment by mfb** [comment]
Meanwhile you could also give Memsource.com a try - it seems to have better XLIFF support (hides the XLIFF markup). I haven't tried other cloud translation services yet.

**Comment by ghost** [comment]
I did try memsource.com after going through its invasive account registration process (it required me to enter my phone number and my real name for some reason), but its editor is even worse than that of Transifex. Here's a screenshot:

![screen shot 2015-07-31 at 16 59 24](https://cloud.githubusercontent.com/assets/3340067/9008862/da91f7a0-37a5-11e5-8e6a-9d1749888326.png)

**Comment by mfb** [comment]
Interesting, when I use Memsource it hides the markup with little numeric icons: 
![memsource](https://cloud.githubusercontent.com/assets/3278/9009300/0cc130d4-3755-11e5-97f4-9868258b57f4.png)

**Comment by ghost** [comment]
Did you tweak any of the settings? I skimmed through the settings but I didn't see anything that could remove the markup.

Edit: I also tried it on my Windows machine (which I don't use for work), but the result remains the same:

![capture](https://cloud.githubusercontent.com/assets/3340067/9009408/9fb8027e-37a9-11e5-8d5a-07cc4183e192.PNG)

Either I'm doing something wrong or we don't have the same file. (I got the file from the repository.)

**Comment by mfb** [comment]
It looks like in your case, for some reason Memsource is using line breaks to detect the strings rather than parsing and interpreting the XLIFF.  What was File Type set to when you imported the file?

**Comment by ghost** [comment]
I could check, but I get 404 error when I try to reach memsource's login page right now. (What the hell?)

But the file I uploaded on memsource is still on my disk and it's an .xlf file, so it should be the same.

**Comment by ghost** [comment]
So, I have an update.

I contacted Transifex's support and asked them why the strings look so messy on their editor. I got this reply just now:

> Hello ,
> 
> Thank you for using Transifex and for submitting your request.
> 
> This is a well known issue and a strange one to solve. 
> We follow XLIFF 1.0 Specification, where
> 
> < and &
> 
> must always be used in place of the character they reference. However this is not the whole case, as we change all the entities once the file is being uploaded on our server. The issue starts to become worrying when someone downloads the translations where the entities reference are in place of the characters. But, there are some developers who prefer the entities reference form the characters.
> 
> Hope I shed some light in the issue.
> 
> Kind regards,
> ~~Name redacted~~

Their response didn't make any sense to me, but that's because I know nothing about XLIFF files or how does one create them. I hope it does help you.

**Comment by mfb** [comment]
Yes transifex doesn't interpret the XLIFF and CDATA markup for whatever reason.. they told me that they haven't had a lot of requests for better XLIFF support (yet)? I think you should try to get the file interpreted as XLIFF by memsource, which I assume wasn't happening when you uploaded there.

**Comment by ghost** [comment]
I'm pretty sure I uploaded the file as .xlf on memsource but their website doesn't give me whole lot of confidence. (I couldn't login their site for 2 straight days.) Which is kind of important considering I'll be doing my work using their website. I'll probably end up buying that app I'd mentioned. That seems like a more secure way to do it. I wouldn't want my work and time go to waste because of some stupid bug on memsource's editor.

I'm curious though, is working on XLIFF files really that necessary? Clearly there isn't good enough support for this file type. Not that I'm complaining but I must have spent at least a few hours trying to find a solution (and I came up with nothing) which is a time I could've spent translating. There may be some people out there who're willing to help, but don't have as much spare time as me to try to find which app or website works the best (or even work at all) with XLIFF files. That's not an ideal situation for both of the involved parties.

Edit: Can you confirm that [these translations](https://github.com/dombili/SMC/blob/master/for_use_eff_secure-messaging-scorecard_tr.xlf) work fine with your system? I only translated small part of the file since there's no guarantee that it'll work. Thanks. (Sorry for creating a different repository, I couldn't upload the file here. I'll remove it after your confirmation.)

**Comment by yurifw** [comment]
Hello all, I'm also having trouble editing XLIFF files, I was about to give up when I e-mailed Kim from EFF and got referred here.

I'm on Linux, can someone recommend me one? The wikipedia page didn't help me, the CAT tools I found keep crashing or refuse to open XLIFF files. Still didn't try memsource, I would rather be able to work offline... 

update:
I'm giving a try to memsource, so far, everything has gone well, but i am also working with xliff files for the first time, so if someone could test [this file](https://github.com/yurifw/www-l10n/blob/master/translations/pt-br/JobID29_en_pt-br.xlf) i would appreciate

**Comment by mfb** [comment]
@yurifw For some reason your translation is missing the XML markup that should be there inside the `<target>` tag. When you were editing in Memsource you should have seen some highlighted tags that you needed to keep there, they will be filled in with HTML when you download the completed translation.

If you want a local tool, on Linux I could recommend Virtaal.  Depending on your distro this might be as easy as `apt-get install virtaal`.  Disabling the plugins under preferences might help if it seems slow (freezing up while you do things).  I found it was slow when editing large strings of content (splitting these up into shorter translatable sections of content might make sense).

**Comment by mfb** [comment]
@dombili You need to start with job 35: https://github.com/EFForg/www-l10n/blob/master/translations/tr/JobID35_en_tr.xlf   Looks like you started with job 21 which is for Farsi.

**Comment by ghost** [comment]
Yeah, that's the file I got from Transifex. It was under the Turkish language section and it was for the Secure Messaging Scorecard page. Strange that it's for Farsi. Anyway. I translated the file you linked to (only small parts of it for testing purposes) and [uploaded here](https://github.com/dombili/Translation-test-/blob/master/JobID35_en_tr.xlf). Please let me know if it works or not. Thanks.

**Comment by yurifw** [comment]
ok, I modified the file, didn't realize they were missing, I think everything is correct now, if you can confirm that, I'll leave you in peace and work on the translation xD

**Comment by mfb** [comment]
@yurifw it worked, although I got a "Failed to validate semantic integrity" warning, probably because only part of the string was there? It's ideal if the XML tags in source and target all match up, so it can be automatically validated and there's no need to eyeball the HTML.

@dombili it worked, although since there are no XML tags (bpt, ept etc.) in your translations, I got a "Failed to validate semantic integrity" warning. It would be _ideal_ if you had a way to include the XML tags in the translation, so the translation can be automatically validated.

**Comment by ghost** [comment]
Unfortunately there isn't a way. My editor looks like [this](https://cloud.githubusercontent.com/assets/3340067/9000815/ff8882fe-3755-11e5-93ae-35dd9a488645.png) (shared previously) and it doesn't even allow me to edit the reference (source) language. I still have to deal with html syntax, but since you don't seem to have an alternative to XLIFF files, I'll just have to suck it up and try to work with it.

If this problem isn't a deal breaker for you, I'll go ahead and buy the app and start translating.

**Comment by obazia** [comment]
Virtaal works fine on Fedora21. I installed it with a simple yum install from the regular repositories and for the first time I'm managing to use a translation tool on Linux without any problems :)

**Comment by mfb** [comment]
@dombili this isn't ideal for me because I can't do automated import. I think it'd be worth trying two things: 1) Virtaal has a Mac beta download and 2) Memsource - did you try setting the file type to XLIFF? And maybe email them if you can't login, or try using a VPN to connect - it could be some network issue?  There are probably other options out there, but Virtaal and Memsource are two  I know of which allow you to export the bpt and ept XLIFF tags, which give us completely automated import.   But if you do use this, we can make it work.. Thanks!

**Comment by ghost** [comment]
I realise this isn't ideal. I was mostly asking whether you can work with/around it. I'm out of my comfort zone as well since I'm not using the tools I'm comfortable with.

1- I installed Virtaal but I got this message on my first run without even seeing the UI:

<img width="632" alt="screen shot 2015-08-05 at 20 54 25" src="https://cloud.githubusercontent.com/assets/3340067/9093345/4af482b2-3bb4-11e5-8ad3-eed532627351.png">

Latest update to its beta version was in 2012, so it's safe to assume Virtaal isn't a viable option. At least for me.

2- I was able to login to memsource after deleting cookies and cache files, but that's not my problem with them. Right now I want, but in the future I need, to be able to work offline. Memsource doesn't allow this. I tried installing their app assuming it'd allow me to work offline but when I try to install their Mac app I get this error message:

<img width="690" alt="screen shot 2015-08-05 at 21 01 25" src="https://cloud.githubusercontent.com/assets/3340067/9093604/acbb7e1e-3bb5-11e5-9933-e0b8c570e230.png">

Along with this, my login problems and their online editor looking bootstrap-y, I don't have a lot of confidence that my translations will be safe there. I could've eased my mind by keeping a local copy of my translations on my computer and solved this issue, as well as the problem you're probably gonna have with automation, but as I said, working online won't be an option for me soon.

I realise I'm looking like a whiny person here, but I really am trying to find a good solution for both of us. It just doesn't seem to exist.

I'll buy [the app I'd mentioned](https://itunes.apple.com/us/app/counterparts-lite/id711576477?mt=12) and start working with that. If there's anything I can do to help you with the automation inconvenience I'll cause you, please let me know.

**Comment by mfb** [comment]
Ok. If you can find a way to "transfer from source" or "copy source to target", preserving the XLIFF markup, that would be great. If not, maybe you can file a feature request :)

**Comment by yurifw** [comment]
@mfb I think you got that warning because I only translated part of the file, and I only inserted the html tags on the translated parts, (in memsource, I started translating from line 359, so before that, there is neither text nor html tags before line 359 in the translation). Could that be it?

**Comment by mfb** [comment]
@yurifw hope so, it's not especially easy to debug :)  Just wanted to flag it. On memsource what I did was click "copy source to target" (Control-Insert) to get all the XLIFF tags in there.
